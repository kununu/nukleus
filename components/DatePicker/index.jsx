import React, {Component, PropTypes} from 'react';
import DatePicker from 'react-datepicker';
import moment from 'moment';

import 'react-datepicker/dist/react-datepicker.css';

import styles from './index.scss';

import {
  errorStyles,
  controlLabel,
  formControl,
  formGroup
} from '../index.scss';

export default class DatePickerComponent extends Component {
  static propTypes = {
    disabled: PropTypes.bool,
    error: PropTypes.string,
    id: PropTypes.string.isRequired,
    inputStyle: PropTypes.string,
    isClearable: PropTypes.bool,
    name: PropTypes.string.isRequired,
    placeholder: PropTypes.string,
    query: PropTypes.object,
    required: PropTypes.bool,
    title: PropTypes.string.isRequired,
    value: PropTypes.string
  };

  static defaultProps = {
    disabled: false,
    error: null,
    inputStyle: 'inline',
    isClearable: true,
    placeholder: '',
    query: {},
    required: false
  };

  state = {
    showError: false,
    value: this.props.value
  }

  componentWillMount () {
    this.updateValue(
      this.props.query[this.props.name] ||
      this.props.value ||
      ''
    );
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.error) this.showError();
    this.updateValue(
      nextProps.query[this.props.name] ||
      nextProps.value ||
      ''
    );
  }

  // Property initializer binds method to class instance
  onChange = date => {
    this.updateValue(date);
    this.hideError();
  };

  updateValue (value) {
    this.setState({value});
  }

  showError () {
    this.setState({showError: true});
  }

  hideError () {
    this.setState({showError: false});
  }

  render () {
    const {
      title,
      name,
      id,
      error,
      inputStyle,
      disabled,
      isClearable,
      placeholder,
      required
    } = this.props;

    return (
      <div className={`${formGroup} ${styles[inputStyle]} ${styles.datePickerContainer}`}>
        <label
          className={controlLabel}
          htmlFor={id}>

          {title}
        </label>

        <div className={styles.innerContainer}>
          <DatePicker
            className={formControl}
            name={name}
            id={id}
            disabled={disabled}
            placeholderText={placeholder}
            selected={this.state.value ? moment(this.state.value) : null}
            isClearable={isClearable}
            showYearDropdown
            required={required}
            onChange={this.onChange} />

          <span className={styles.icon}>
            <i className="fa fa-calendar" aria-hidden="true" />
          </span>
        </div>

        {this.state.showError &&
          <span className={errorStyles}>
            {error}
          </span>}
      </div>
    );
  }
}
