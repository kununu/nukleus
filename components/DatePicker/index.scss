@import '~kununu-theme/scss/base/variables';

.datePickerContainer {
  width: 100%;

  * {
    box-sizing: border-box;
  }

  input {
    cursor: pointer;

    + a {
      position: absolute;
      right: 30px;
      width: 100%;

      &:after {
        background-color: $gray-base-80;
        margin-top: -30px;
        padding-right: 3px;
        padding-top: 0;
        font-size: $font-size-base;
        right: 10px;
      }
    }
  }
}

.innerContainer {
  position: relative;

  > div {
    width: 100%;
  }
}

.icon {
  position: absolute;
  z-index: 1;
  font-size: $font-size-lg;
  color: $gray-base-60;
  top: 9px;
  right: 12px;
  pointer-events: none;
}

.container {
  position: relative;
  width: 100%;
}

.inline {
  display: flex;
  flex-flow: row;
  align-items: center;

  @media (min-width: $screen-sm-min) {
    padding-top: 6px;
  }

  label {
    margin-right: 10px;
    width: 140px;
    margin-bottom: 0;
    text-align: left;

    + div {
      width: 100%;
    }
  }

  @media (max-width: $screen-xs-min) {
    display: block;
  }

  &.paddingTop {
    @media (min-width: $screen-sm-min) {
      padding-top: 25px;
    }
  }
}
