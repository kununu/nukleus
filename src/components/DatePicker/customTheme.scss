@import '~@kununu/kununu-theme-v2/scss/base/variables';

.innerContainer {
  position: relative;
}

.formGroup {
  position: relative;
  margin-bottom: 6px;
  width: 80%;
}

.formControl {
  box-shadow: inset 0 1px 1px rgba($black, .075);
  transition: border-color ease-in-out .15s, box-shadow ease-in-out .15s;
  display: block;
  width: 100%;
  min-height: 60px;
  padding: 0 10px;
  font-size: $font-size-sm;
  color: $blue;
  background-color: $input-bg;
  background-image: none; // Reset unusual Firefox-on-Android default style; see https://github.com/necolas/normalize.css/issues/214
  border: 1px solid $gray-base-20;
  border-radius: $input-border-radius; // Note: This
}

.icon {
  position: absolute;
  right: 20px;
  padding-top: 15px;
  font-size: 20px;
}

