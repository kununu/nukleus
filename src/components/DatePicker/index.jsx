import React from 'react';
import PropTypes from 'prop-types';
import DatePicker from 'react-datepicker';

import 'react-datepicker/dist/react-datepicker.css';

import styles from './index.scss';

import Error from '../Error';
import {
  controlLabel,
  controlLabelRequired,
  controlNote,
  formControl,
  formControlError,
  formGroup,
  srOnly
} from '../index.scss';

export default class DatePickerComponent extends React.Component {
  static propTypes = {
    dateFormat: PropTypes.string,
    disabled: PropTypes.bool,
    error: PropTypes.string,
    errorSubInfo: PropTypes.string,
    icon: PropTypes.element,
    id: PropTypes.string.isRequired,
    inputStyle: PropTypes.string,
    isRequired: PropTypes.bool,
    label: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.object
    ]),
    labelHidden: PropTypes.bool,
    name: PropTypes.string.isRequired,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    query: PropTypes.object,
    requiredLabel: PropTypes.string,
    title: PropTypes.string,
    value: PropTypes.string
  };

  static defaultProps = {
    dateFormat: 'dd/MM/yyyy',
    disabled: false,
    error: null,
    errorSubInfo: null,
    icon: null,
    inputStyle: 'inline',
    isRequired: false,
    label: null,
    labelHidden: false,
    onBlur: () => {},
    onChange: () => {},
    onFocus: () => {},
    query: {},
    requiredLabel: '',
    title: null,
    value: ''
  };

  state = {
    showError: false,
    value: this.props.value
  }

  componentWillMount () {
    const {
      query,
      name,
      value
    } = this.props;

    this.updateValue(query[name] || value || '');

    // Show error, if already set
    if (this.props.error !== null) this.showError();
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.error) this.showError();
    this.updateValue(nextProps.query[this.props.name] ||
      nextProps.value ||
      '');
  }

  // Property initializer binds method to class instance
  onChange = value => {
    const {
      onChange,
      name
    } = this.props;

    this.updateValue(value);
    this.hideError();

    /*
     * react-datepicker won't return a native DOM event
     * on their onChange callback, so we need to create
     * a fake one, because kununu-form-wrapper and other
     * form libs will expect a DOM event object to process this field
     */
    onChange({
      target: {
        name,
        value
      }
    });
  };

  get label () {
    const {
      id,
      title,
      label,
      labelHidden
    } = this.props;

    if (!label && !title) return null;

    if (labelHidden) return <span className={srOnly}>{label || title}</span>;

    /**
     * generates the TextField label based on the Textfield label prop
     *
     * @return {ReactElement} [Either returns a label or a react element with the added css class labelContainer]
    */
    if (typeof label === 'string' || typeof title === 'string') {
      return <label className={`${controlLabel} ${this.hasError() ? styles.controlLabelError : ''}`} htmlFor={id}>{label || title}</label>;
    }

    // We don't simply put a more complex element inside a label to prevent a
    // clickable element like a link or button inside a label
    // However to also add the labelContainer class, we need to return a cloned
    // element and not just the label - element itself
    return React.cloneElement(
      label,
      {
        ...label.props,
        className: controlLabel
      }
    );
  }

  showError () {
    this.setState({showError: true});
  }

  hideError () {
    this.setState({showError: false});
  }

  updateValue (value) {
    this.setState({value});
  }

  hasError () {
    return this.state.showError && this.props.error;
  }

  render () {
    const {
      name,
      icon,
      id,
      error,
      errorSubInfo,
      inputStyle,
      disabled,
      isRequired,
      requiredLabel,
      dateFormat
    } = this.props;

    return (
      <div className={`${formGroup} ${styles[inputStyle]} ${styles.datePickerContainer} ${requiredLabel ? styles.paddingTop : ''}`} id={`${name}-container`}>
        {requiredLabel &&
          <span className={`${controlNote} ${controlLabelRequired}`}>
            {requiredLabel}
          </span>
        }

        {this.label}

        <div className={styles.innerContainer}>
          <DatePicker
            className={`${formControl} ${this.hasError() ? formControlError : ''}`}
            name={name}
            id={id}
            disabled={disabled}
            dateFormat={dateFormat}
            selected={this.state.value ? this.state.value : null}
            required={isRequired}
            onBlur={this.props.onBlur}
            onChange={this.onChange}
            onFocus={this.props.onFocus} />
          {icon ?
            <span className={styles.icon}>
              {icon}
            </span>
            : ''}

          {this.hasError() &&
            <Error
              info={error}
              subInfo={errorSubInfo} />
          }
        </div>

      </div>
    );
  }
}
