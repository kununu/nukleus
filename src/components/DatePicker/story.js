import React from 'react';
import {storiesOf} from '@storybook/react';
import {defaultOptions} from 'defaultOptions';
import {withOptions} from '@storybook/addon-options';
import {action} from '@storybook/addon-actions';
import {withInfo} from '@storybook/addon-info';
import {withKnobs, text, boolean} from '@storybook/addon-knobs/react';

import DatePicker from './index';

storiesOf('DatePicker', module)
  .addDecorator(withOptions({...defaultOptions}))
  .addDecorator(withKnobs)
  .addDecorator((story, context) => withInfo('Basic datepicker')(story)(context))
  .add('basics', () => (
    <DatePicker
      dateFormat={text('dateFormat', 'dd/MM/yyyy')}
      icon={(
        <i
          className="fa fa-calendar"
          aria-hidden="true"
        />
)}
      id="date-picker"
      inputStyle="block"
      isRequired={boolean('isRequired', false)}
      label="DatePicker"
      labelHidden={boolean('labelHidden', false)}
      name="date-picker"
      onBlur={action('onBlur')}
      onChange={action('onChange')}
      onChangeRaw={action('onChangeRaw')}
      onClick={action('onClick')}
      onFocus={action('onFocus')}
      requiredLabel={text('requiredLabel', 'required')}
      showAbbreviatedMonthDropdown={boolean('showAbbreviatedMonthDropdown', false)}
      showMonthDropdown={boolean('showMonthDropdown', false)}
      showYearDropdown={boolean('showYearDropdown', false)}
      value={text('value', new Date())}
    />
  ));
