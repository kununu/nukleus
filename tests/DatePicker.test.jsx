import React from 'react';
import renderer from 'react-test-renderer';
import DatePicker from 'DatePicker'; // eslint-disable-line import/no-unresolved, import/extensions, import/no-extraneous-dependencies

jest.mock('react-datepicker', () => 'Datepicker');
jest.mock('moment', () => 'Moment');

test('Renders datepicker without crashing', () => {
  const component = renderer.create(
    <DatePicker
      id="date-picker"
      isRequired
      name="date-picker"
      title="DatePicker" />
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
