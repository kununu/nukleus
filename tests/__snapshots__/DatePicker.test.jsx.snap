exports[`test Renders datepicker without crashing 1`] = `
<div
  className="formGroup inline datePickerContainer">
  <label
    className="controlLabel"
    htmlFor="date-picker">
    DatePicker
  </label>
  <div
    className="innerContainer">
    <Datepicker
      className="formControl"
      disabled={false}
      id="date-picker"
      isClearable={true}
      name="date-picker"
      onChange={[Function]}
      placeholderText=""
      required={false}
      selected={null}
      showYearDropdown={true} />
    <span
      className="icon">
      <i
        aria-hidden="true"
        className="fa fa-calendar" />
    </span>
  </div>
</div>
`;
